### ISSUE TITLE
Issues with Tauristar

### ISSUE THREAD

**Issue by mbabker** [Digital-Peak-Incubator/tpl_tauristar#105]
- The template breaks parameter injection by reading the template configuration from `JFactory`, the injected parameters in the `JDocument` instance should be used instead
- The template makes use of deprecated method calls; as this is new code this already sets a bad precedent for things to come
- The template's CSS framework is being added as a global dependency; Joomla 3.x is Bootstrap 2 based, introducing Bootstrap 3 media to the global directory is just going to create a :swear:storm.  Put your media in the template.

**Comment by zero-24** [comment]
>The template makes use of deprecated method calls; as this is new code this already sets a bad precedent for things to come

What deprected code? Cou you point us to that code?

>The template breaks parameter injection by reading the template configuration from JFactory, the injected parameters in the JDocument instance should be used instead

I'm not sure that i understand you correct? Can you point me to the code that do that so we can fix that?

**Comment by laoneo** [comment]
> The template's CSS framework is being added as a global dependency; Joomla 3.x is Bootstrap 2 based, introducing Bootstrap 3 media to the global directory is just going to create a :swear:storm. Put your media in the template.

In theory it is BS2, but in reality almost all templates do come with BS3. What we try to achieve here is to disconnect the core from the frontend framework as described in our proposal https://docs.google.com/document/d/1GQXDG2A_Xq8fgjXgiMuzC9hnQzPVUX46d2pogNdLP4s/edit?usp=sharing. If we get that then we can do Joomla 5 much faster and don't have to wait for another 5 years to release it. If the site admins need to create a new template when they upgrade to Joomla 4, then it is not a one click upgrade and we have the same problem as from 1 to 2 to 3.

**Comment by mbabker** [comment]
Except it is perfectly possible to take a 2.5 template and run it just fine
on 3.x.  I did it with my own site until I had time to properly rebuild on
a newer template.

Once all framework related decisions are at the template level and the core
layouts are truly agnostic, then this vision of needing to ship all these
framework related files goes away.  They can be shoved into the template
where they belong and the assets in the media directory becomes aspects of
the Joomla core only (its JavaScript API, extension specific styling rules
(preferably decoupled from the templates to make those easier to re-use and
override too), and imagery used as part of the core APIs).

As things are today, all of the default layouts are truly assets of the
default templates (Isis and Protostar, not the legacy templates).  The
default layouts (those in the extension's tmpl directory) should NOT be
designed against a specific template.  However, it makes things easier for
templates to adopt the existing layouts and makes it easier for extension
developers to skin layouts to fit into a wide variety of templates by
designing them a specific way, so the mentality is the heart of Joomla's
templating is Bootstrap 2 based when that is 100% not the case (the old
help screen CMS extension and template used zero Bootstrap or jQuery
elements, so it is indeed possible to build without loading these "core"
dependencies).

**Comment by laoneo** [comment]
That's why we have here the approach to load all the framework related code into their own files like default.bs3.php as the pr https://github.com/Digital-Peak-Incubator/tpl_tauristar/pull/99/files shows (The include still needs to be done better and will be replaced on some point). Like that we keep bc and do move away the fw code from the default.php files.

**Comment by mbabker** [comment]
That PR is pure freakin' crap, I'm not even going to sugarcoat it.  For starters it has introduced a hardcoded dependency to a layout file and cannot be overridden, next it introduces the notion of broadcasting frameworks into the codebase, and as I've already explained there is no logical reason to do this.

Unsubscribing, all continuing to discuss things here is going to do is get my blood pressure high.

**Comment by laoneo** [comment]
As I said, the include is just a temporary solution. It shows one approach. If you have a better solution to disconnect the core from a frontend framework, then we are interested to hear your opinion. Just saying Joomla 3 comes with BS2 and Joomla 4 with BS4 without a proper migration sounds like we have not learned to lessons from the past.

Hope your blood pressure comes down again.

**Comment by mbabker** [comment]
The approach is simple.  The core UI (the default layouts used by the core extensions and the default `JLayout` files) should be blissfully framework unaware.  All of the framework specific stuff belongs in the template and ONLY the template.  The only issue with this approach is it makes it a bit harder to quickly plug in templates using the core layouts (you pretty much have to use and maintain a larger set of layout overrides).

There is no need for `JLayout` to have a parameter to load framework specific files.  There is no need for `JHtmlBootstrap` to have a version switch.  There is no need for a `framework` parameter in your templates.  Use the system that we have in place today.

**Comment by laoneo** [comment]
I totally agree with you on that, there should be no framework code in the default layouts, but it is today. 

With your approach you copy files and have duplicated content. Is this really your solution for the problem? Extension developers will not like that.

**Comment by mbabker** [comment]
Do you think extension developers are going to like being told they need to support at least two template frameworks in their extension layout code?  Or all the conditionals that are going to come with that?  You're going to force devs to go back to the practice of different extension code bases per Joomla version because you're going to make the layout handling code too complex to do it all in one.  Or we're going to force devs to only support one Joomla version at a time meaning those still running the "LTS" version are going to lose support quickly or those who are adopting the new version are going to do so without the extensions they need.

Do you honestly believe that the Joomla community has the resources to make core properly support multiple CSS frameworks?  We have enough issues today getting efficient JavaScript written or patches that involve testing parts of the application stack that aren't something the user can apply a patch via patch tester and click a couple of things in the backend UI tested through.

Adding parameters all over the place to broadcast a specific framework (optionally a specific version of a specific framework) is only going to make everything in Joomla harder.  From writing distributed extensions to creating simple layout overrides, everything becomes more complex when you add that parameter/toggle.

**Comment by laoneo** [comment]
> Do you think extension developers are going to like being told they need to support at least two template frameworks in their extension layout code?

No, but it is reality today and it will be the case that we have to support 3 frameworks when Joomla 4 comes out with bs4. The other way is that the extension does load it's own namespaced version of bootstrap as many do nowadays, which is bad as well.
All framework related code should be extracted from the core files. With this approach we give the extension devs a way to have it as painless as possible. I agree, that it adds a new layer of complexity, but at least it brings in more opinions and possible solutions.

> Adding parameters all over the place to broadcast a specific framework

We are not going to do that. Why do you think so?

> everything becomes more complex when you add that parameter/toggle

This hidden parameter your are refering is just an example place to illustrate that the template is defining which framework it is built against.

Joomla core needs to be disconnected from any frontend framework. We try to find here a solution for that and appreciate every input.
